File: src/main/java/org/khun/leetcode/TwoPointers/ValidPalindrome.java
package org.khun.leetcode.TwoPointers;

public class ValidPalindrome {

    public static boolean isPalindrome(String s) {
        String s1 = s.replaceAll("[^a-zA-Z0-9]+","").toLowerCase();
        StringBuilder sb = new StringBuilder(s1.trim());
        sb.reverse();
        return s1.equals(sb.toString());
    }

    public static void main(String[] args) {
        System.out.println(isPalindrome("A man, a plan, a canal: Panama"));
        System.out.println(isPalindrome("race a car"));
        System.out.println(isPalindrome(" "));
    }
}

File: src/main/java/org/khun/leetcode/TwoPointers/ValidPalindrome.kt
package org.khun.leetcode.TwoPointers

fun isPalindrome(s: String): Boolean {
    if (s.isNullOrEmpty()) return true;
    val str = s.lowercase().replace(Regex("[^a-z0-9]"), "")
    var left = 0
    var right = str.length - 1
    while (left < right) {
        if (str[left] != str[right]) return false
        left++
        right--
    }
    return true
}

fun main() {
    println(isPalindrome("A man, a plan, a canal: Panama"))
    println(isPalindrome("race a car"))
    println(isPalindrome(" "))
}

